services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusDataTest.kt

```diff
@@ -24,293 +32,790 @@ import org.junit.jupiter.params.provider.NullSource
+                                                        geometry = GeometriaFactory.thirdPolygon()
+                                                    ),
+                                                    ApplicationFactory.createTyoalue(
+                                                        geometry = GeometriaFactory.thirdPolygon()
+                                                    ),
+                                                    ApplicationFactory.createTyoalue(
+                                                        geometry = GeometriaFactory.secondPolygon()
+                                                    ),
+                                                )
+                                        )
+                                    )
+                            )
+
+                        val result = base.listChanges(dataWithTwoWorkAreas)
+
+                        assertThat(result)
+                            .containsExactly(
+                                "areas[0]",
+                                "areas[0].tyoalueet[1]",
+                                "areas[0].tyoalueet[1].geometry",
+                                "areas[0].tyoalueet[2]",
+                                "areas[0].tyoalueet[3]",
+                            )
+                    }
+
+                    @Test
+                    fun `returns several changes when a work area is added to the middle`() {
+                        val updated =
+                            base.copy(
+                                areas =
+                                    listOf(
+                                        ApplicationFactory.createExcavationNotificationArea(
+                                            tyoalueet =
+                                                listOf(
+                                                    ApplicationFactory.createTyoalue(
+                                                        geometry = GeometriaFactory.polygon()
+                                                    ),
+                                                    ApplicationFactory.createTyoalue(
+                                                        geometry = GeometriaFactory.thirdPolygon()
+                                                    ),
+                                                    ApplicationFactory.createTyoalue(
+                                                        geometry = GeometriaFactory.thirdPolygon()
+                                                    ),
+                                                    ApplicationFactory.createTyoalue(
+                                                        geometry = GeometriaFactory.secondPolygon()
+                                                    ),
+                                                )
+                                        )
+                                    )
+                            )
```

Comment:
The deep nesting makes these really hard to read. Some intermediary values would make these a lot more readable.
```suggestion
                    val tyoalueet =
                        listOf(
                            ApplicationFactory.createTyoalue(geometry = GeometriaFactory.polygon()),
                            ApplicationFactory.createTyoalue(
                                geometry = GeometriaFactory.thirdPolygon()
                            ),
                            ApplicationFactory.createTyoalue(
                                geometry = GeometriaFactory.thirdPolygon()
                            ),
                            ApplicationFactory.createTyoalue(
                                geometry = GeometriaFactory.secondPolygon()
                            ),
                        )
                    val area =
                        ApplicationFactory.createExcavationNotificationArea(tyoalueet = tyoalueet)
                    val updated = base.copy(areas = listOf(area))
```

Reply:
Changed as suggested.